Add guide for LLM eval (#10058)
mlflow/mlflow: docs/source/llm-eval/llm-eval.rst

> .. _llm-eval:

Reviewer: @chenmoneygithub @prithvikannan Can we base this off the https://github.com/mlflow/mlflow/tree/docs-refactor branch and include it in the LLM guide? 

@BenWilson2 can you review the guide from a best practices perspective?

Reviewer 2: This doc can be moved to `docs/source/llms/llm-evaluate/index.rst` and it will serve as the 'landing page' to llm evaluation. I'll leave more comments below about structure.

> ====================================

Reviewer 2: Can we stick to the standard rst mapping?

<h1> -> =====
<h2> -> ------
<h3> -> ^^^^^
<h4> -> """"""

And we really shouldn't go beyong h4 (even though up to h6 is supported, it looks silly)

MLflow LLM Evaluate
====================================

With the emerging of ChatGPT, LLMs have shown its power of text generation in various fields, such as 
question answering, translating and text summarization. Evaluating LLMs' performance is slightly different 
from traditional ML models, as very often there is no single ground truth to compare against. 
MLflow provides an API :py:func:`mlflow.evaluate()` to help evaluate your LLMs.

> ``mlflow.evaluate()`` consists of 3 main components:

Reviewer 2: ```suggestion
MLflow's LLM evaluation functionality consists of 3 main components:
```

1. **Metrics**: the metrics to compute, LLM evaluate will use LLM metrics. 
2. **A model to evaluate**: it can be an MLflow ``pyfunc`` model, a URI pointing to one registered 

>    MLflow model, or any python callable that represents your model, e.g, a HuggingFace text summarization pipeline. 

Author: done!

3. **Evaluation data**: the data your model is evaluated at, it can be a pandas Dataframe, a python list, a 
   numpy array or an :py:func:`mlflow.data.dataset.Dataset` instance.

Quickstart
==========

Below is a simple example that gives an quick overview of how MLflow LLM evaluation works. The example builds
a simple question-answering model by wrapping "openai/gpt-3.5-turbo" with custom prompt. You can paste it to
your IPython or local editor and execute it, and install missing dependencies as prompted. Running the code 
requires OpenAI API key, if you don't have an OpenAI key, you can set it up [here](https://platform.openai.com/account/api-keys).

.. code-block:: python

    import mlflow
    import openai
    import os
    import pandas as pd
    from getpass import getpass

    # Set up openai auth.
    OPENAI_API_KEY = getpass()
    os.environ["OPENAI_API_KEY"] = OPENAI_API_KEY

>     openai.api_key = OPENAI_API_KEY

Reviewer 2: Could we just show a bash command of setting the env var for the open ai key?

>     openai.api_key = OPENAI_API_KEY

Author: @BenWilson2 It has some error, for some reason we have to set up both. 

@prithvikannan Ideally we shouldn't require users to set up both env var and `openai.api_key`, and users may not set env var before doing `import openai`. Are we fixing the bug that both env var and `openai.api_key` are required?

>     openai.api_key = OPENAI_API_KEY

Reviewer 3: we have filed https://github.com/mlflow/mlflow/issues/10021 as a good first issue.

>     openai.api_key = OPENAI_API_KEY

Author: I see, in that case I would keep the code as it is, so users could always copy & paste the workable code regardless of their setup.

        # Wrap "gpt-3.5-turbo" as an MLflow model.
        logged_model_info = mlflow.openai.log_model(
            model="gpt-3.5-turbo",
            task=openai.ChatCompletion,
            artifact_path="model",
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": "{question}"},
            ],
        )

        # Use predefined question-answering metrics to evaluate our model.
        results = mlflow.evaluate(
            logged_model_info.model_uri,
            eval_data,
            targets="ground_truth",
            model_type="question-answering",
        )
        print(f"See aggregated evaluation results below: \n{results.metrics}")

        # Evaluation result for each data record is available in `results.tables`.
        eval_table = results.tables["eval_results_table"]
        print(f"See evaluation table below: \n{eval_table}")

LLM Evaluation Metrics
=======================

There are two types of LLM evaluation metrics in MLflow:

1. Metrics relying on SaaS model (e.g., OpenAI) with prompt engineering, e.g., :py:func:`mlflow.metrics.relevance`, as called 

>    LLM-as-judge metrics. These  metrics are created via :py:func:`make_genai_metric` method. For each data record, these metrics 

Author: sg, let me delete that

   under the hood sends one prompt consisting of the following information to the SaaS model, and extract the score from model response:

   * Metrics definition.
   * Metrics grading criteria.
   * Reference examples.
   * Input data/context.
   * Model output.
   * [optional] Ground truth.

   More details of how these fields are set can be found in the section "Create your Custom LLM-evaluation Metrics".

> 

Author: yes, I have updated the doc with pre-canned metrics explanation.

2. Function-based per-row metrics. These metrics calculate a score for each data record (row in terms of Pandas/Spark dataframe),
   based on certain functions, like Rouge (:py:func:`mlflow.metrics.rougeL`) or Flesch Kincaid (:py:func:`mlflow.metrics.flesch_kincaid_grade_level`). 
   These metrics are similar to traditional metrics.

Select Metrics to Evaluate
--------------------------

There are two ways to select metrics to evaluate your LLMs:

1. Use ``model_type`` argument in ``mlflow.evaluate``, every predefined model type comes with a 
   set of metrics. This is suitable if your model falls in our predefined categories, e.g., 
   ``question-answering``. You can find the full list of model types, and the associated evaluate 

>    metrics [here](https://mlflow.org/docs/latest/python_api/mlflow.html#mlflow.evaluate). A sample code is as below:

Reviewer 2: I just looked at that list from the API docs and it isn't particularly user-friendly. We should list these out as part of this guide.

>    metrics [here](https://mlflow.org/docs/latest/python_api/mlflow.html#mlflow.evaluate). A sample code is as below:

Reviewer 2: What about something like this?

```rst

MLflow LLM evaluation includes several default evaluation collections of metrics for specific tasks. Depending on the type of LLM use case that you 
are interested in evaluating, these pre-defined collections can greatly simplify the process of running evaluations. 

The supported metrics for given model types are shown below:

* **question-answering**:

    * exact-match
    * `mean_perplexity <https://huggingface.co/spaces/evaluate-metric/perplexity>`_ :sup:`1`
    * `toxicity_ratio <https://huggingface.co/spaces/evaluate-measurement/toxicity>`_ :sup:`1`
    * `mean_ari_grade_level <https://en.wikipedia.org/wiki/Automated_readability_index>`_ :sup:`2`
    * `mean_flesch_kincaid_grade_level <https://en.wikipedia.org/wiki/Flesch%E2%80%93Kincaid_readability_tests#Flesch%E2%80%93Kincaid_grade_level>`_ :sup:`2`

* **text-summarization**: 

    * `ROUGE <https://huggingface.co/spaces/evaluate-metric/rouge>`_ :sup:`3`
    * `mean_perplexity <https://huggingface.co/spaces/evaluate-metric/perplexity>`_ :sup:`1`
    * `toxicity_ratio <https://huggingface.co/spaces/evaluate-measurement/toxicity>`_ :sup:`1`
    * `mean_ari_grade_level <https://en.wikipedia.org/wiki/Automated_readability_index>`_ :sup:`2`
    * `mean_flesch_kincaid_grade_level <https://en.wikipedia.org/wiki/Flesch%E2%80%93Kincaid_readability_tests#Flesch%E2%80%93Kincaid_grade_level>`_ :sup:`2`

* **text models**:

    * `mean_perplexity <https://huggingface.co/spaces/evaluate-metric/perplexity>`_ :sup:`1`
    * `toxicity_ratio <https://huggingface.co/spaces/evaluate-measurement/toxicity>`_ :sup:`1`
    * `mean_ari_grade_level <https://en.wikipedia.org/wiki/Automated_readability_index>`_ :sup:`2`
    * `mean_flesch_kincaid_grade_level <https://en.wikipedia.org/wiki/Flesch%E2%80%93Kincaid_readability_tests#Flesch%E2%80%93Kincaid_grade_level>`_ :sup:`2`

:sup:`1` Requires `evaluate <https://pypi.org/project/evaluate>`_, `pytorch <https://pytorch.org/get-started/locally/>`_, and 
`transformers <https://huggingface.co/docs/transformers/installation>`_

:sup:`2` Requires `textstat <https://pypi.org/project/textstat>`_

:sup:`3` Requires `evaluate <https://pypi.org/project/evaluate>`_, `nltk <https://pypi.org/project/nltk>`_, and 
`rouge-score <https://pypi.org/project/rouge-score>`_

However, using the pre-defined metrics associated with a given model type is not the only way to generate scoring metrics for LLM evaluation in MLFlow. 
There are two methods for generating evaluation metrics for your LLMs with MLflow:

1. Specify the ``model_type`` argument in :py:func:`mlflow.evaluate` 

    * Each predefined model type comes with a standard set of metrics that are available for relevant evaluation of a model type. 
    * The defaults are suitable if your model falls in one of the predefined categories (e.g., ``question-answering``).   

An example of using the predefined metrics for a given ``model_type`` is shown below:

.. code-section:: 
    .. code-block:: python

        results = mlflow.evaluate(
            model,
            eval_data,
            targets="ground_truth",
            model_type="question-answering",
        )

2. Specify a custom list of metrics by explicitly referencing a metric calculation function.

    * To add additional metrics to the default collection from part 1 above, add the function names to the ``extra_metrics`` argument.
    * To diable default metric calculation and only calculate explicit metrics, remove the ``model_type`` argument and define the desired metrics. 

An example of disabling the default metrics and explicitly declaring a subset of metrics to calculate is shown below:

.. code-section::
    .. code-block:: python

        results = mlflow.evaluate(
            model,
            eval_data,
            targets="ground_truth",
            extra_metrics=[mlflow.metrics.toxicity(), mlflow.metrics.latency()],
        )

The full reference for supported evaluation metrics can be found `here <../python_api/mlflow.html#mlflow.evaluate>`_. 

```

>    metrics [here](https://mlflow.org/docs/latest/python_api/mlflow.html#mlflow.evaluate). A sample code is as below:

Author: applied!

>     .. code-block:: python

Reviewer 2: Please wrap all code-block entries with a 

```text
.. code-section::
    .. code-block:: python
```
so that we get an indication of language

Author: Thanks! I will do it as a followup since without `.. code-section::` gives a better rendering in preview, which makes code reviewer easy.

        results = mlflow.evaluate(
            model,
            eval_data,
            targets="ground_truth",
            model_type="question-answering",
        )

2. Use ``extra_metrics`` argument in ``mlflow.evaluate``, you can add your required metrics to the list, 
   and also include your custom metrics (will be covered in the next section). Notice that if both 
   ``model_type`` and ``extra_metrics`` are set, ``extra_metrics`` together with predefined metrics are computed. 

>    In order to just use your handpicked metrics, please remove ``model_type``. A sample code is as below:

Reviewer 2: ```suggestion
   In order to just use your handpicked metrics, please remove ``model_type``, as shown in the example below:
```

2. Pass ``eval_fn`` and other arguments to ``mlflow.metricsmake_metric`` API to create the metric. 

The following code creates a dummy per-row metric called ``"over_10_chars"``: if the model output is greater than 10, 
the score is 1 otherwise 0.

.. code-block:: python

    def eval_fn(predictions, targets, metrics):
        scores = []
        for i in range(len(predictions)):
            if len(predictions[i]) > 10:
                scores.append(1)
            else:
                scores.append(0)
        return MetricValue(
            scores=scores,
            aggregate_results=standard_aggregations(scores),
        )

    # Create an EvaluationMetric object.
    passing_code_metric = make_metric(
        eval_fn=eval_fn, greater_is_better=False, name="over_10_chars"
    )

Prepare Your LLM for Evaluating
=====================================

In order to evaluate your LLM with ``mlflow.evaluate()``, your LLM has to be one of the following type:

1. A :py:func:`mlflow.pyfunc.PyFuncModel` instance or a URI pointing to a logged `mlflow.pyfunc.PyFuncModel` model. In
   general we call that MLflow model.

> 2. A python function that takes in string inputs and outputs a single string. 

Author: Yea I saw that requirement, but feel confused about what it means. for example the following works as a custom callable for evaluate:
```
def openai_qa(inputs):
    answers = []
    system_prompt = "Please answer the following question in formal language."
    for index, row in inputs.iterrows():
        completion = openai.ChatCompletion.create(
            model="gpt-3.5-turbo",
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": "{row}"},
            ],
        )
        answers.append(completion.choices[0].message.content)

    return answers
```
And it just returns a python list?

> 2. A python function that takes in string inputs and outputs a single string. 

Reviewer 3: yea, that works because its list -> list, which is a valid predict fn. see https://mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#mlflow.pyfunc.PyFuncModel.predict

> 2. A python function that takes in string inputs and outputs a single string. 

Author: got it, let me add it
